intellij-plugin-structure/structure-base/src/main/kotlin/com/jetbrains/plugin/structure/base/utils/CompatibilityUtils.kt

```diff
@@ -0,0 +1,61 @@
+package com.jetbrains.plugin.structure.base.utils
+
+import java.util.ArrayList
+
+
+class CompatibilityUtils {
+  companion object {
+    private const val MAX_BRANCH_VALUE = 1000
+    private const val MAX_COMPONENT_VALUE = 10000
+    private const val MAX_BUILD_VALUE = 100000
+    private val NUMBERS_OF_NINES by lazy { initNumberOfNines() }
+
+    fun versionAsLong(vararg components: Int): Long {
+      val baselineVersion = components.getOrElse(0) { 0 }
+      val build = components.getOrElse(1) { 0 }
+      var longVersion = branchBuildAsLong(baselineVersion, build)
+
+      if (components.size >= 3) {
+        val component = components[2]
+        longVersion += if (component == Integer.MAX_VALUE) MAX_COMPONENT_VALUE - 1 else component
+      }
+
+      return longVersion
+    }
+
+    fun getMaxVersionAsLong(): Long {
+      return branchBuildAsLong(MAX_BRANCH_VALUE - 1, MAX_BUILD_VALUE - 1)
+    }
+
+    private fun branchBuildAsLong(branch: Int, build: Int): Long {
+      val result = if (build == Integer.MAX_VALUE || isNumberOfNines(build)) {
+        MAX_BUILD_VALUE - 1
+      } else {
+        build
+      }
+
+      return branch.toLong() * MAX_COMPONENT_VALUE * MAX_BUILD_VALUE + result.toLong() * MAX_COMPONENT_VALUE
+    }
+
+    private fun isNumberOfNines(p: Int) = NUMBERS_OF_NINES.any { it == p }
+
+    private fun initNumberOfNines(): IntArray {
+      val numbersOfNines = ArrayList<Int>()
+      var i = 99999
```

Comment:
Why `99999` is a starting point? Why not `999` or  `9999`?

Reply:
Because `MAX_BUILD_VALUE = 10000`. That means that we will use five digits for the second component in `version as long.` We use this approach to represent Idea versions as long. So, we want to reuse it with Ktor. Since we are talking about lower bounds, I think that it's not a problem. Are you agree?

Reply:
Ok, sounds good